convert2rhel/backup.py

```diff
@@ -0,0 +1,201 @@
+            if critical:
+                loggerinst.critical("Error: Couldn't install %s packages." % pkgs_as_str)
+
+            loggerinst.warning("Couldn't install %s packages." % pkgs_as_str)
+            return False
+
+        for path in pkgs_to_install:
+            nvra, _ = os.path.splitext(os.path.basename(path))
+            self.track_installed_pkg(nvra)
+
+        return True
+
+    def restore_pkgs(self):
+        """Restore system to the original state."""
+        self._remove_installed_pkgs()
+        remove_orphan_folders()
+        self._install_removed_pkgs()
+
+
+class RestorableFile(object):
+    def __init__(self, filepath):
+        self.filepath = filepath
+
+    def backup(self):
+        """Save current version of a file"""
+        loggerinst.info("Backing up %s." % self.filepath)
+        if os.path.isfile(self.filepath):
+            try:
+                loggerinst.debug("Copying %s to %s." % (self.filepath, BACKUP_DIR))
+                shutil.copy2(self.filepath, BACKUP_DIR)
+            except (OSError, IOError) as err:
+                # IOError for py2 and OSError for py3
+                loggerinst.critical("Error(%s): %s" % (err.errno, err.strerror))
+        else:
+            loggerinst.info("Can't find %s.", self.filepath)
+
+    def restore(self):
+        """Restore a previously backed up file"""
+        backup_filepath = os.path.join(BACKUP_DIR, os.path.basename(self.filepath))
+        loggerinst.task("Rollback: Restoring %s from backup" % self.filepath)
+
+        if not os.path.isfile(backup_filepath):
+            loggerinst.warning("%s hasn't been backed up" % self.filepath)
+            return
+        try:
+            shutil.copy2(backup_filepath, self.filepath)
+        except (OSError, IOError) as err:
+            # Do not call 'critical' which would halt the program. We are in
+            # a rollback phase now and we want to rollback as much as possible.
+            # IOError for py2 and OSError for py3
+            loggerinst.warning("Error(%s): %s" % (err.errno, err.strerror))
+            return
+        loggerinst.info("File %s restored" % self.filepath)
+
+
+class RestorablePackage(object):
+    def __init__(self, pkgname):
+        self.name = pkgname
+        self.path = None
+
+    def backup(self):
+        """Save version of RPM package"""
+        loggerinst.info("Backing up %s" % self.name)
+        if os.path.isdir(BACKUP_DIR):
+            reposdir = get_hardcoded_repofiles_dir()
+
+            if reposdir and system_info.corresponds_to_rhel_eus_release():
+                # When backing up the packages, the original system repofiles are still available and for them we can't
+                # use the releasever for RHEL repositories
+                self.path = download_pkg(
+                    self.name,
+                    dest=BACKUP_DIR,
+                    set_releasever=False,
+                    reposdir=reposdir,
+                )
```

Comment:
```suggestion
            if reposdir and system_info.corresponds_to_rhel_eus_release():
                # One of the reasons we hardcode repofiles pointing to archived repositories of older system minor versions is that we need to be able to download an older package version as a backup. Because for example the default repofiles on CentOS Linux 8.4 point only to 8.latest repositories that already don't contain 8.4 packages.
                # The reason for the EUS condition is that convert2rhel allows conversions of just those older system minor versions that correspond to a RHEL EUS release.
                self.path = download_pkg(
                    self.name,
                    dest=BACKUP_DIR,
                    # When backing up the packages, the original system repofiles are still available and for them we can't
                    # use the releasever for RHEL repositories
                    set_releasever=False,
                    reposdir=reposdir,
                )
```

Reply:
done.